src/mem/ruby/common/WriteMask.cc

```diff
@@ -55,5 +55,27 @@ WriteMask::print(std::ostream& out) const
         << std::flush;
 }
 
+void
+WriteMask::performAtomic(uint8_t * p,
+        std::deque<uint8_t*>& log) const
+{
+    int offset;
+    uint8_t *block_update;
+    // Here, operations occur in FIFO order from the mAtomicOp
+    // vector. This is done to match the ordering of packets
+    // that was seen when the initial coalesced request was created.
+    for (int i = 0; i < mAtomicOp.size(); i++) {
+        // Save the old value of the data block in case a
+        // return value is needed
+        block_update = new uint8_t[mSize];
+        std::memcpy(block_update, p, mSize);
+        log.push_back(block_update);
```

Comment:
I am wondering whether you could push most of the bookeping to the AtomicOpFunctor....

Reply:
I think it could be done. To do this though, the response messages would have to carry the WriteMask back up the hierarchy, since the vector of functors are held by the WriteMask itself. Since the DataBlock is already a part of the response message, I opted to put the logging structure there. Would it make more sense to use the AtomicOpFunctors? I think putting it there might make it a little clunky: in the GPUCoalescer, we would read from the return WriteMasks atomicVector of functors instead of the from the datablock.

Reply:
> I think it could be done. To do this though, the response messages would have to carry the WriteMask back up the hierarchy, since the vector of functors are held by the WriteMask itself.
Since the DataBlock is already a part of the response message, I opted to put the logging structure there. Would it make more sense to use the AtomicOpFunctors? I think putting it there might make it a little clunky: in the GPUCoalescer, we would read from the return WriteMasks atomicVector of functors instead of the from the datablock.

Apologies for the late reply, I was on a small break from work. Yeah I understand, and binding the log of changes to the block itself makes a lot of sense. My original message was just meant to move the three lines to the "prologue" of the functor to keep the WriteMask cleaner

        block_update = new uint8_t[mSize];
        std::memcpy(block_update, p, mSize);
        log.push_back(block_update);

But I realise it is a little save...